torchaudio/datasets/utils.py

```diff
@@ -0,0 +1,341 @@
+        if key.startswith("download_warning"):
+            return value
+
+    return None
+
+
+def _save_response_content(response, destination, chunk_size=32768):
+    with open(destination, "wb") as f:
+        pbar = tqdm(total=None)
+        progress = 0
+        for chunk in response.iter_content(chunk_size):
+            if chunk:  # filter out keep-alive new chunks
+                f.write(chunk)
+                progress += len(chunk)
+                pbar.update(progress - pbar.n)
+        pbar.close()
+
+
+def _is_tar(filename):
+    return filename.endswith(".tar")
+
+
+def _is_targz(filename):
+    return filename.endswith(".tar.gz")
+
+
+def _is_gzip(filename):
+    return filename.endswith(".gz") and not filename.endswith(".tar.gz")
+
+
+def _is_zip(filename):
+    return filename.endswith(".zip")
+
+
+def extract_archive(from_path, to_path=None, remove_finished=False):
+    if to_path is None:
+        to_path = os.path.dirname(from_path)
+
+    if _is_tar(from_path):
+        with tarfile.open(from_path, "r") as tar:
+            tar.extractall(path=to_path)
+    elif _is_targz(from_path):
+        with tarfile.open(from_path, "r:gz") as tar:
+            tar.extractall(path=to_path)
+    elif _is_gzip(from_path):
+        to_path = os.path.join(
+            to_path, os.path.splitext(os.path.basename(from_path))[0]
+        )
+        with open(to_path, "wb") as out_f, gzip.GzipFile(from_path) as zip_f:
+            out_f.write(zip_f.read())
+    elif _is_zip(from_path):
+        with zipfile.ZipFile(from_path, "r") as z:
+            z.extractall(to_path)
+    else:
+        raise ValueError("Extraction of {} not supported".format(from_path))
+
+    if remove_finished:
+        os.remove(from_path)
+
+
+class Cache:
+    """
+    Wrap a generator so that, whenever a new item is returned, it is saved to disk in a pickle.
+    """
+
+    def __init__(self, generator, location):
+        self.generator = generator
+        self.location = location
+
+        self._id = id(self)
+        self._cache = []
+        self._internal_index = 0
+
+    def __iter__(self):
+        self._internal_index = 0
+        return self
+
+    def __next__(self):
+        if self._internal_index < len(self):
+            item = self[self._internal_index]
+        else:
+            item = next(self.generator)
+
+            file = str(self._id) + "-" + str(len(self))
+            file = os.path.join(self.location, file)
+            self._cache.append(file)
+
+            makedir_exist_ok(self.location)
+            with open(file, "wb") as file:
+                pickle.dump(item, file)
+
+        self._internal_index += 1
+        return item
+
+    def __getitem__(self, index):
+        file = self._cache[index]
+        with open(file, "rb") as file:
+            item = pickle.load(file)
+        return item
+
+    def __len__(self):
+        # Return length of cache
+        return len(self._cache)
+
+
+class Buffer:
```

Comment:
For this to provide a performance gain it will probably want to make use of multithreading. This does introduce various constraints on the given generator that need to be determined. I imagine this to be most useful if applied to function that perform IO operations such as reading from multiple files, downloading multiple URLs etc.

See the following snippet

```
import time
import concurrent.futures
import urllib.request

def read_50(fn):
    for _ in range(50):
        yield fn()

def read_fn():
    with urllib.request.urlopen(<your_image_url>) as response:
        return response.read()

def read_50_threaded(fn):
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        futures = [executor.submit(fn) for _ in range(50)]
        for future in futures:
            yield future.result()

t1 = time.time()
results = list(read_50(read_fn))
t2 = time.time()

print("t: " + str(t2 - t1))

t1 = time.time()
results = list(read_50_threaded(read_fn))
t2 = time.time()

print("t: " + str(t2 - t1))
```